test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +279,698 @@ func TestAdoption(t *testing.T) {
+	pods = getPods(t, podClient, labelMap)
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %d, want 1", len(pods.Items))
+	}
+	// Verify deleting pod survives
+	if pods.Items[0].Name != deletingPod.Name {
+		t.Fatalf("expected deleting pod %s survives, but it is not found", pods.Items[0].Name)
+	}
+	// Also verify its deletion timestamp is not nil
+	if pods.Items[0].DeletionTimestamp == nil {
+		t.Fatalf("expected deletion timestamp of deleting pod %s is not nil, but it is nil", pods.Items[0].Name)
+	}
+
+	// Then scale up RS back to 2 replicas, and verify 3 pods exist:
+	// deleting pod, 2 new pods replacing deleting and failed pods
+	scaleRS(t, c, rs, 2)
+	pods = getPods(t, podClient, labelMap)
+	if len(pods.Items) != 3 {
+		t.Fatalf("len(pods) = %d, want 3", len(pods.Items))
+	}
+	foundDeletingPod := false
+	foundFailedPod := false
+	for _, pod := range pods.Items {
+		if pod.Name == deletingPod.Name {
+			foundDeletingPod = true
+		}
+		if pod.Name == failedPod.Name {
+			foundFailedPod = true
+		}
+	}
+	// Verify (again) deleting pod survives; not need to verify its deletion
+	// timestamp is not nil (irrelevant in proving creation of 2 new pods)
+	if !foundDeletingPod {
+		t.Fatalf("expected deleting pod %s survives, but it is not found", deletingPod.Name)
+	}
+	// Verify failed pod is deleted
+	if foundFailedPod {
+		t.Fatalf("expected failed pod %s does not exist, but it is found", failedPod.Name)
+	}
+	// There are 3 pods: one of them is deleting pod, and failed pod is not among the others
+	// It implies that 2 new pods MUST have been created to replace the deleting and failed pods
+	// Hence, no verification needed
 
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":"1"}},"template":{"metadata":{"labels":{"uniqueKey":"1"}}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	// Fetch RS again before changing its annotations
+	rsClient := c.Extensions().ReplicaSets(ns.Name)
+	rs, err := rsClient.Get(rs.Name, metav1.GetOptions{})
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+
+	// Add an annotation to pod template to test RS's status does update without replicas change
+	rs.Annotations = map[string]string{"test": "annotation"}
+	newRS, err := rsClient.Update(rs)
+	if err != nil {
+		t.Fatalf("failed to update rs %s: %v", rs.Name, err)
+	}
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		curRS, err := rsClient.Get(rs.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return curRS.Status.ObservedGeneration >= newRS.Generation, nil
```

Comment:
@enisoc I think `.Status.ObservedGeneration` has to be strictly greater than `.Generation`.